kernel/src/hil/accel.rs

```diff
@@ -0,0 +1,52 @@
+//! Interface for Accelerators
+
+use crate::common::leasable_buffer::LeasableBuffer;
+use crate::ErrorCode;
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a, const T: usize> {
+    /// This callback is called when the binary data has been loaded
+    /// On error or success `input` will contain a reference to the original
+    /// data supplied to `load_binary()`.
+    fn binary_load_done(&'a self, result: Result<(), ErrorCode>, input: &'static mut [u8]);
+
+    /// This callback is called when a operation is computed.
+    /// On error or success `output` will contain a reference to the original
+    /// data supplied to `run()`.
+    fn op_done(&'a self, result: Result<(), ErrorCode>, output: &'static mut [u8; T]);
+}
+
+/// A generic accellerator. This can be used to accelerate any type of
+/// operation.
+pub trait Accel<'a, const T: usize> {
+    /// Set the client instance which will receive
+    fn set_client(&'a self, client: &'a dyn Client<'a, T>);
+
+    /// Load the acceleration binary data into the accelerator.
+    /// This data will be accelerator specific and could be an
+    /// elf file which will be run or could be binary settings used to
+    /// configure the accelerator.
+    /// This function can be called multiple times if multiple binary blobs
+    /// are required.
+    /// There is no guarantee the data has been written until the `binary_load_done()`
+    /// callback is fired.
+    /// On error the return value will contain a return code and the original data
+    fn load_binary(
+        &self,
+        input: LeasableBuffer<'static, u8>,
+    ) -> Result<(), (ErrorCode, &'static mut [u8])>;
+
+    /// Run the acceleration operation.
+    /// This doesn't return any data, instead the client needs to have
+    /// set a `op_done` handler to determine when this is complete.
+    /// On error the return value will contain a return code and the original data
+    /// If there is data from the `load_binary()` command asyncrously waiting to
+    /// be written it will be written before the operation starts.
+    fn run(&'a self, output: &'static mut [u8; T])
+        -> Result<(), (ErrorCode, &'static mut [u8; T])>;
```

Comment:
Why pass a sized array instead of a slice? It seems possible some accelerators would want to support variable length inputs

Reply:
This is an output, the idea is that the output is fixed by the hardware

Reply:
My mistake. Still, it seems possible some accelerators would size their outputs based on the size of the input

Reply:
My worry with an output buffer is what if a user passes a too small buffer.

For example if the output of the accelerator is 64 bytes, but a user passes a `[u8; 32]` we will throw a `NOMEM` error (instead of only half filling it), but then the user has no way of knowing how big the buffer really should be. My expectation is that the output buffer size of a accelerator is constant (although the useful information will change based on protocol).

The other way of handling it could be just to trust the caller and if they pass a smaller buffer just fill the small buffer as they might not want the entire output. That seems prone to errors though.

If I change it to a slice what should we do for length mismatches?

Reply:
That's a reasonable point. If its a slice I think that returning an Error on length mismatches would be fine -- it could be up to the caller to pass an appropriate length buffer. I suppose it is fine to leave it as fixed size until we have a concrete example of an accelerator that does not provide a fixed size output